### PR TITLE
[WIP] Create GCE deployment script

### LINKED ISSUE #1246: Create Google Container Engine deployment script
# Short Description

A script that could deploy loklak to Google Cloud from the cloud shell would be helpful in easy deployment. This script would 
* Clone the project
* Modify certain variables in the `conf/config.properties`
* Build and push container image
* Create and expose deployment

### PR DESCRIPTION
**Pull request by singhpratyush**
### Short description

Fixes #1246.

Usage:
```sh
$ wget http://link.for/script/to/deploy.py
$ python deploy.py -h
usage: deploy.py [-h] [-q] [-n] [-u USER] [-b BRANCH] -p PROJECT

optional arguments:
  -h, --help            show this help message and exit
  -q, --quite           Quite mode
  -n, --new             Create a new cluster
  -u USER, --user USER  Username of Github repo owner
  -b BRANCH, --branch BRANCH
                        Branch to deploy
  -p PROJECT, --project PROJECT
                        GCP project ID
```

Created instance at http://104.154.174.213/ using the script.

#### Defaults

| Property | JSON Key | Value | Details |
|-----------|-------------|----------|----------|
| Version of build | `version` | `v1` | The version of tagging the registry for GCR |
| Cluster name | `cluster` | `loklak-cluster` | Name of Kubernetes cluster |
| Number of nodes in cluster | `num-nodes` | `2` | Number of nodes while creating a cluster |
| Node type | `node-type` | `n1-standard-4` | Node type, as supported by GCP |
| Deploy zone | `zone` | `us-central1-a` | Zone where the cluster has been deployed, or where you wish to deploy |
| Deployment name | `deployment` | `loklak` | Name of Kubernetes deployment |

#### Sample run - 
```bash
$ python deploy.py -u singhpratyush -b 1246 -p loklak-201451036 -n
INFO: Cloning https://github.com/singhpratyush/loklak_server.git...
INFO: Successfully cloned https://github.com/singhpratyush/loklak_server.git- 
Note: checking out 'origin/1246'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:
  git checkout -b <new-branch-name>
HEAD is now at 31f0aeb6... Fix #1246: Create GCE deployment script
INFO: 
INFO: Loaded configuration: {u'node-type': u'n1-standard-8', u'config-changes': {u'retrieval.forbackend.loops': u'30', u'Xmx': u'20G', u'retrieval.forbackend.concurrency': u'5'}, u'num-nodes': u'1'}
INFO: Creating Cluster...
Creating cluster loklak-cluster...done.                                                                                                                                                                                                    
Created [https://container.googleapis.com/v1/projects/loklak-201451036/zones/us-central1-a/clusters/loklak-cluster].
kubeconfig entry generated for loklak-cluster.
INFO: Cluster created successfully - 
NAME            ZONE           MASTER_VERSION  MASTER_IP        MACHINE_TYPE   NODE_VERSION  NUM_NODES  STATUS
loklak-cluster  us-central1-a  1.6.4           104.154.170.179  n1-standard-8  1.6.4         1          RUNNING
INFO: Changing property retrieval.forbackend.loops to 30
INFO: Changing property Xmx to 20G
INFO: Changing property retrieval.forbackend.concurrency to 5
INFO: Building image with TAG gcr.io/loklak-201451036/loklak:v1
INFO: Docker build complete - 
Sending build context to Docker daemon 252.6 MB
Step 1 : FROM alpine:latest
 ---> a41a7446062d
Step 2 : MAINTAINER Ansgar Schmidt <ansgar.schmidt@gmx.net>
 ---> Using cache
 ---> 8f9f856e0015
Step 3 : ENV LANG en_US.UTF-8
 ---> Using cache
 ---> 36307540d939
Step 4 : EXPOSE 80 443
 ---> Using cache
 ---> 241e3c514599
Step 5 : ADD bin /loklak_server/bin/
 ---> Using cache
 ---> 365ead99c5f9
Step 6 : ADD conf /loklak_server/conf/
 ---> Using cache
 ---> eb3d06035662
Step 7 : ADD src /loklak_server/src/
 ---> Using cache
 ---> 3f66363a1f87
Step 8 : ADD html /loklak_server/html/
 ---> Using cache
 ---> d3002d003598
Step 9 : ADD installation /loklak_server/installation/
 ---> Using cache
 ---> dee790f8c22c
Step 10 : ADD ssi /loklak_server/ssi/
 ---> Using cache
 ---> 8cee43efaa0c
Step 11 : ADD gradle /loklak_server/gradle/
 ---> Using cache
 ---> 057a215b8cab
Step 12 : ADD gradlew /loklak_server/
 ---> Using cache
 ---> 4657cf28904f
Step 13 : ADD build.gradle /loklak_server/
 ---> Using cache
 ---> 5cf739cd4b2d
Step 14 : ADD settings.gradle /loklak_server/
 ---> Using cache
 ---> 28afd21f774c
Step 15 : RUN apk update && apk add openjdk8 git bash &&     cd /loklak_server && ./gradlew build &&     sed -i 's/^\(port.http=\).*/\180/;s/^\(port.https=\).*/\1443/;s/^\(upgradeInterval=\).*/\186400000000/'         conf/config.propert
ies &&     apk del openjdk8 git &&     apk add openjdk8-jre-base
 ---> Using cache
 ---> 8c4c39b87ffe
Step 16 : WORKDIR /loklak_server
 ---> Using cache
 ---> 580918e8bea4
Step 17 : VOLUME /loklak_server/data
 ---> Using cache
 ---> b2e50fd7611c
Step 18 : CMD /loklak_server/bin/start.sh -Idn
 ---> Using cache
 ---> a63870e4bf55
Successfully built a63870e4bf55

INFO: Pushing docker image to GCR...
INFO: Pushed gcr.io/loklak-201451036/loklak:v1 to GCR - 
The push refers to a repository [gcr.io/loklak-201451036/loklak]
bf7fd9bbf537: Preparing
d25bd8644db8: Preparing
31adb0d04b7e: Preparing
a15f370156ea: Preparing
11d09eacc36e: Preparing
4cef0a3995ef: Preparing
6e7353a40a23: Preparing
c9566b9cf68d: Preparing
e2a86d8d781d: Preparing
f82330e234b4: Preparing
01110dd4a2ac: Preparing
3fb66f713c9f: Preparing
4cef0a3995ef: Waiting
6e7353a40a23: Waiting
c9566b9cf68d: Waiting
e2a86d8d781d: Waiting
f82330e234b4: Waiting
01110dd4a2ac: Waiting
3fb66f713c9f: Waiting
bf7fd9bbf537: Layer already exists
11d09eacc36e: Layer already exists
a15f370156ea: Layer already exists
6e7353a40a23: Layer already exists
4cef0a3995ef: Layer already exists
31adb0d04b7e: Layer already exists
d25bd8644db8: Layer already exists
c9566b9cf68d: Layer already exists
e2a86d8d781d: Layer already exists
01110dd4a2ac: Layer already exists
3fb66f713c9f: Layer already exists
f82330e234b4: Layer already exists
v1: digest: sha256:e8aed2580cfbf47b9f1be7dbc8aa88c00e4451cb48faa14f74ff12b34019e73a size: 2791
INFO: Running deployment loklak using tag gcr.io/loklak-201451036/loklak:v1 
INFO: deployment "loklak" created
INFO: Exposing deployment loklak
INFO: service "loklak" exposed
INFO: Polling to get public IP for loklak
104.154.174.213
```

#### TODO
- [x] Allow modifications in `conf/config.properties`
- [x] Use the config file already present in the repository after cloning.
- [x] Handle `CTRL+C` gracefully.
- [ ] ...

I have:
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

### CODE REVIEW

**Comment by vibhcool** [comment]
Wouldn't it be better to write shell script?
instead of writing python script and using subprocess.

**Comment by singhpratyush** [comment]
@vibhcool: Initially I was planning to go with a `bash` script too. But a `python` script would be easy to maintain. We can have simple argument processing, `JSON` parsing, etc. with `python`. And with the breakdown of different tasks, it would become much simple to tune in future.

File: google-cloud-kubernetes/deploy.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+def expose_deployment(deployment):
+    info('Exposing deployment %s' % deployment)
+    output = subprocess.check_output(['kubectl', 'expose', 'deployment',
+                                      deployment, '--type=LoadBalancer'])
+    info(output)
+
+
+def fetch_public_ip(deployment):
+    info('Polling to get public IP for loklak')
+    regex_ip = re.compile(r'%s[\t\ ]+[0-9]+\.[0-9]+\.[0-9]+\.[0-9]'
```
Comment (vibhcool) [inline google-cloud-kubernetes/deploy.py]: IMHO, using API here instead of scraping would be better solution, eventhough we are scrapers 😂

Comment (singhpratyush) [inline google-cloud-kubernetes/deploy.py]: @vibhcool: I'm sorry but I didn't get the API that you are talking about. Could you please share a link or something?

**Comment by djmgit** [comment]
@singhpratyush looks like codacy is not happy with subprocess :/

File: google-cloud-kubernetes/deploy.py (hunk 1 above)
Comment (vibhcool) [inline google-cloud-kubernetes/deploy.py]: https://cloud.google.com/container-engine/reference/rest/

**Comment by singhpratyush** [comment]
@djmgit: Added keyboard interrupt to **TODO**. Thanks for the suggestion.

**Comment by singhpratyush** [comment]
@Achint08 @djmgit @hemantjadon @kavithaenair @SarishiNoHara @SKrPl @vibhcool: The script is now functional and tested. Please review.

Also, try to deploy using the script and see if you face any issues.

File: google-cloud-kubernetes/config.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"config-changes":{"Xmx":"20G","retrieval.forbackend.concurrency":"5","retrieval.forbackend.loops":"30"},"num-nodes":"1","node-type":"n1-standard-8"}
```
Comment (kavithaenair) [inline google-cloud-kubernetes/config.json]: What if user wants different configurations? Change the configurations in the file?

Comment (singhpratyush) [inline google-cloud-kubernetes/config.json]: Yes, this file will be used after cloning the repo and checking out the branch.

**Comment by singhpratyush** [comment]
@djmgit: Codacy can be configured to ignore such issues. But it can only be done by the corresponding user.

File: google-cloud-kubernetes/config.json (hunk 2 above)
Comment (niranjan94) [inline google-cloud-kubernetes/config.json]: This isn't the right way to specify config. 

- No of nodes and type should move to a kubernetes config file
- App config should move to a configmap and referenced while creating a pod

**Comment by niranjan94** [comment]
And also, I have to agree with @vibhcool ... Python script just seems like an overkill when much complex things can be done via a simple bash file and kubernetes & gcloud cli if used properly.

**Comment by singhpratyush** [comment]
Thanks, @niranjan94 for the guidance. I'll modify the PR accordingly.

**Comment by niranjan94** [comment]
Great @singhpratyush :) Lemme know if you need an help regarding this

**Comment by singhpratyush** [comment]
Closing for now. Will work on this later.

**Pull request closed without merging**